beam/core/src/main/java/cz/o2/proxima/beam/direct/io/CommitLogRead.java

```diff
@@ -0,0 +1,194 @@
+        String name,
+        long limit,
+        RepositoryFactory repositoryFactory,
+        CommitLogReader.Factory<?> readerFactory) {
+
+      this.name = name;
+      this.limit = limit;
+      this.repositoryFactory = repositoryFactory;
+      this.readerFactory = readerFactory;
+    }
+
+    @ProcessElement
+    public ProcessContinuation process(
+        RestrictionTracker<OffsetRestrictionTracker.StartEndOffset, Offset> tracker,
+        OutputReceiver<StreamElement> output) {
+
+      try (ObserveHandle handle = startObserve(tracker.currentRestriction())) {
+        int i = 0;
+        while (!Thread.currentThread().isInterrupted()
+            && observer.getWatermark() < Watermarks.MAX_WATERMARK) {
+
+          StreamElement element = observer.take();
+          if (element == null) {
+            ExceptionUtils.ignoringInterrupted(() -> TimeUnit.MILLISECONDS.sleep(100));
+            continue;
+          }
+          boolean isFirst = i++ == 0;
+          Offset offset = Objects.requireNonNull(observer.getLastReadContext().getOffset());
+          if (externalizableOffsets
+              && !tracker.currentRestriction().isStartInclusive()
+              && isFirst) {
+            // skip the exclusive first offset
+            continue;
+          }
+          if (!tracker.tryClaim(offset)) {
+            return ProcessContinuation.stop();
+          }
+          output.outputWithTimestamp(element, Instant.ofEpochMilli(element.getStamp()));
+        }
+      }
+      boolean terminated = observer.getWatermark() >= Watermarks.MAX_WATERMARK;
+      return terminated
+          ? ProcessContinuation.stop()
+          : ProcessContinuation.resume().withResumeDelay(Duration.millis(100));
+    }
+
+    private ObserveHandle startObserve(OffsetRestrictionTracker.StartEndOffset restriction) {
+      CommitLogReader reader = readerFactory.apply(repositoryFactory.apply());
+      this.externalizableOffsets = reader.hasExternalizableOffsets();
+      observer = BlockingQueueLogObserver.create(name, limit, Watermarks.MIN_WATERMARK);
+      return reader.observeBulkOffsets(
+          Collections.singletonList(restriction.getStartOffset()), observer);
+    }
+
+    @GetInitialRestriction
+    public OffsetRestrictionTracker.StartEndOffset initialRestriction() {
+      return OffsetRestrictionTracker.StartEndOffset.initialRestriction();
+    }
+
+    @SplitRestriction
+    public void splitRestriction(
+        @Restriction OffsetRestrictionTracker.StartEndOffset restriction,
+        OutputReceiver<OffsetRestrictionTracker.StartEndOffset> splits) {
+
+      if (restriction.isInitial()) {
+        CommitLogReader reader = readerFactory.apply(repositoryFactory.apply());
+        // compute starting offsets from commit log reader
+        List<Partition> partitions = reader.getPartitions();
+        try (ObserveHandle handle =
+            reader.observeBulkPartitions(partitions, Position.CURRENT, droppingObserver())) {
```

Comment:
position should be configurable

Reply:
The position is ignored here.

Reply:
how comes? If I understand this correctly it is used for fetching starting offsets, which is now always point to "current" position?

Reply:
Ah, my bad. You're right, I will pull this parameter to constructor of the PTransform.